nexus/src/db/queries/network_interface.rs

```diff
@@ -979,16 +989,342 @@ impl QueryFragment<Pg> for InsertNetworkInterfaceQueryValues {
         out.push_identifier(dsl::ip::NAME)?;
         out.push_sql(", ");
         out.push_identifier(dsl::slot::NAME)?;
+        out.push_sql(", ");
+        out.push_identifier(dsl::is_primary::NAME)?;
         out.push_sql(") ");
         self.0.walk_ast(out)
     }
 }
 
+/// A small helper subquery that automatically assigns the `is_primary` column
+/// for a new network interface.
+///
+/// An instance with any network interfaces must have exactly one primary. (An
+/// instance may have zero interfaces, however.) This subquery is used to insert
+/// the value `true` if there are no extant interfaces on an instance, or
+/// `false` if there are.
+#[derive(Debug, Clone, Copy)]
+struct IsPrimaryNic {
+    instance_id: Uuid,
+}
+
+impl QueryId for IsPrimaryNic {
+    type QueryId = ();
+    const HAS_STATIC_QUERY_ID: bool = false;
+}
+
+impl QueryFragment<Pg> for IsPrimaryNic {
```

Comment:
(No change requested) I continue to be awestruck how you've been able to construct complex CTEs by breaking them down into simple subqueries, without imposing complex trait bounds.

That balance between power and readability has been, and continues to be, really impressive!

Reply:
That's very kind of you Sean, thanks!